File: Task 3/C++Stock Viewer/Product/Tech.cpp
#include "Tech.h"

Tech::Tech() : Electronics(){
    type = "";
}

Tech::Tech(std::string name, float cost, std::string supplier, float discount_amount, int packing_units, float cost_per_unit, std::string manufacturer, std::string type) 
            : Electronics(name, cost, supplier, discount_amount, packing_units, cost_per_unit, manufacturer){
    this->type = type;
};

Tech::Tech(const Tech& other) : Electronics(other) {
    this->type = other.type;
}

std::string Tech::display_product_info(){
    return Electronics::display_product_info()+
            "\nType: "+type;
}

// ========================================== PROTOBUFFERS
void Tech::copyToProto(PBStockItem* message){
    PBTech* pbTech = message->mutable_tech();
    pbTech->set_type(type);

    PBElectronics* pbElectronics = pbTech->mutable_electronics();
    pbElectronics->set_manufacturer(manufacturer);

    PBProduct* pbProduct = pbElectronics->mutable_product();
    pbProduct->set_name(name);
    pbProduct->set_cost(cost);
    pbProduct->set_supplier(supplier);

    PBPackage* pbPackage = pbProduct->mutable_packaging();
    pbPackage->set_units(packaging.getUnits());
    pbPackage->set_costperunit(packaging.getCost_Per_Unit());
}

void Tech::copyFromProto(const PBStockItem& message) {
    const PBTech& pbTech = message.tech();
    name = pbTech.electronics().product().name();
    cost = pbTech.electronics().product().cost();
    supplier = pbTech.electronics().product().supplier();
    packaging.setUnits(pbTech.electronics().product().packaging().units());
    packaging.setCost_Per_Unit(pbTech.electronics().product().packaging().costperunit());
    manufacturer = pbTech.electronics().manufacturer();
    type = pbTech.type();
}

File: Task 3/Java/logistics/src/main/java/com/logistics/Components/Product/Tech.java
package com.logistics.Components.Product;

import java.util.Calendar;

import com.logistics.Serialization.PBLogistics.PBElectronics;
import com.logistics.Serialization.PBLogistics.PBPackage;
import com.logistics.Serialization.PBLogistics.PBProduct;
import com.logistics.Serialization.PBLogistics.PBStockItem;
import com.logistics.Serialization.PBLogistics.PBTech;

public class Tech implements Electronics {
    private float cost;
    private String name;
    private String supplier;
    private float discountAmount;
    private Packaging packaging;
    private String manufacturer;
    private String type;

    public Tech() {
        cost = 0;
        name = "";
        supplier = "";
        discountAmount = 1;
        packaging = new Packaging(0, 0);
        manufacturer = "";
        type = "";
    }

    public Tech(String name, float cost, String supplier, float discountAmount, int packingUnits, float costPerUnit, String manufacturer, String type) {
        this.name = name;
        this.cost = cost;
        this.supplier = supplier;
        this.discountAmount = discountAmount;
        this.packaging = new Packaging(packingUnits, costPerUnit);
        this.manufacturer = manufacturer;
        this.type = type;
    }

    // Copy Constructor
    public Tech(Tech other) {
        this.name = other.name;
        this.cost = other.cost;
        this.supplier = other.supplier;
        this.discountAmount = other.discountAmount;
        this.packaging = new Packaging(other.packaging);
        this.manufacturer = other.manufacturer;
        this.type = other.type;
    }

    // Clone method
    @Override public Tech clone() {
        return new Tech(this);
    }

    // Getters
    @Override public float getCost() { return cost; }
    @Override public String getName() { return name; }
    @Override public String getSupplier() { return supplier; }
    @Override public float getDiscountAmount(){return discountAmount;}
    @Override public float getPackagingCost() { return packaging.calculateTotalCost(); }
    @Override public String getManufacturer() { return manufacturer; }
    public String getType() { return type; }

    // Setters
    @Override public void setCost(float cost) { this.cost = cost; }
    @Override public void setName(String name) { this.name = name; }
    @Override public void setSupplier(String supplier) { this.supplier = supplier; }
    @Override public void setDiscountAmount(float discountAmount){this.discountAmount = discountAmount;}
    @Override public void setPackagingData(int packingUnits, float costPerUnit) { this.packaging = new Packaging(packingUnits, costPerUnit); }
    @Override public void setManufacturer(String manufacturer) { this.manufacturer = manufacturer; }
    public void setType(String type) { this.type = type; }

    @Override public float calculateDiscount() {
        // Get the current month
        int currentMonth = Calendar.getInstance().get(Calendar.MONTH);

        // Check if the current month is December 
        if (currentMonth == Calendar.DECEMBER && (discountAmount >= 0 && discountAmount <= 1)) {
            return (1 - discountAmount);
        } else {
            return 1;
        }
    }

    @Override public String display_product_info() {
        return  "\nName: " + name +
                "\nCost: " + cost +
                "\nSupplier: " + supplier +
                "\nDiscount Amount: " + calculateDiscount() + 
                "\nManufacturer: " + manufacturer +
                "\nType: " + type;
    }



    // ========================================== PROTOBUFFERS
    @Override public void copyToProto(PBStockItem.Builder message){
        PBTech.Builder pbTech = PBTech.newBuilder();
        pbTech.setType(type);

        PBElectronics.Builder pbElectronics = pbTech.getElectronicsBuilder();
        pbElectronics.setManufacturer(manufacturer);

        PBProduct.Builder pbProduct = pbElectronics.getProductBuilder();
        pbProduct.setName(name);
        pbProduct.setCost(cost);
        pbProduct.setSupplier(supplier);

        PBPackage.Builder pbPackage = pbProduct.getPackagingBuilder();
        pbPackage.setUnits(packaging.getUnits());
        pbPackage.setCostPerUnit(packaging.getCostPerUnit());

        message.setTech(pbTech);

    };
    @Override public void copyFromProto(PBStockItem message){
        PBTech pbTech = message.getTech();
        name = pbTech.getElectronics().getProduct().getName();
        cost = pbTech.getElectronics().getProduct().getCost();
        supplier = pbTech.getElectronics().getProduct().getSupplier();
        packaging.setUnits(pbTech.getElectronics().getProduct().getPackaging().getUnits());
        packaging.setCostPerUnit(pbTech.getElectronics().getProduct().getPackaging().getCostPerUnit());
        manufacturer = pbTech.getElectronics().getManufacturer();
        type = pbTech.getType();
    };

}
